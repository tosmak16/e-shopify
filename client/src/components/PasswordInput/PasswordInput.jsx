import { useState, useEffect } from 'react';
import zxcvbn from 'zxcvbn';

import TextInput, { INPUT_TYPES } from '../TextInput/TextInput';
import styles from './PasswordInput.scss';

export const PASSWORD_STRENGTH_TYPE = {
  WEAK: 'weak',
  MEDIUM: 'medium',
  STRONG: 'strong'
};

const PasswordInput = props => {
  const [password, setPassword] = useState({
    value: '',
    error: '',
    passwordStrength: ''
  });

  const [passwordVisibility, setPasswordVisibility] = useState(false);

  const validatePassword = (value, placeholder = 'Password') => {
    const { size } = props;
    let error;
    if (!value.length) {
      error = `${placeholder} can not be empty`;
    } else if (value.length < size) {
      error = `${placeholder} should at least be ${size} characters`;
    } else {
      error = '';
    }
    return error;
  };

  const calculatePasswordStrength = value => {
    const results = zxcvbn(value);
    if (results.score <= 1) {
      return PASSWORD_STRENGTH_TYPE.WEAK;
    }
    if (results.score === 2) {
      return PASSWORD_STRENGTH_TYPE.MEDIUM;
    }
    return PASSWORD_STRENGTH_TYPE.STRONG;
  };

  const inputChange = event => {
    const { placeholder, value } = event.target;
    const { onChange } = props;
    const error = validatePassword(value, placeholder);
    const passwordStrength = calculatePasswordStrength(value);
    setPassword({ error, value, passwordStrength });
    return onChange({
      value,
      error,
      isValid: !error.length,
      passwordStrength
    });
  };

  const { value, error, passwordStrength } = password;
  const { showPasswordStrength, shouldShowVisibility, updatedPassword } = props;

  const handlePasswordVisibility = () => {
    setPasswordVisibility(!passwordVisibility);
    props.handleVisibilityStatus(!passwordVisibility);
  };

  useEffect(() => {
    setPassword({
      value: updatedPassword.value,
      error: updatedPassword.error,
      passwordStrength: updatedPassword.passwordStrength
    });
  }, [updatedPassword]);

  return (
    <div>
      <TextInput
        {...props}
        type={passwordVisibility ? INPUT_TYPES.TEXT : INPUT_TYPES.PASSWORD}
        isValid={!error.length}
        onChange={inputChange}
        value={value}
        errorMessage={error}
        onChangePasswordVisibilityStatus={handlePasswordVisibility}
        shouldShowVisibility={shouldShowVisibility}
      />
      {error.length === 0 && value.length > 0 && showPasswordStrength && (
        <span styleName="password-strength-label">
          {/* eslint-disable-next-line react/jsx-one-expression-per-line */}
          Strength:<span styleName={`password-${passwordStrength}`}>{passwordStrength}</span>
        </span>
      )}
    </div>
  );
};

PasswordInput.propTypes = {
  onChange: PropTypes.func.isRequired,
  size: PropTypes.number,
  showPasswordStrength: PropTypes.bool,
  shouldShowVisibility: PropTypes.bool,
  handleVisibilityStatus: PropTypes.func
};

PasswordInput.defaultProps = {
  size: 8,
  showPasswordStrength: false,
  shouldShowVisibility: false,
  handleVisibilityStatus: () => {}
};

export default CSSModules(PasswordInput, styles);
