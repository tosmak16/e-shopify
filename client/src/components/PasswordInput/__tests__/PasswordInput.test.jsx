import PasswordInput from '../PasswordInput';

describe('PasswordInput', () => {
	const component = <PasswordInput onChange={jest.fn()} id="password" />;
	it('renders', () => {
		const tree = shallow(component);

		expect(tree).toMatchSnapshot();
	});

	it('should be able to validate short password', () => {
		const onChange = jest.fn();
		const tree = shallow(<PasswordInput onChange={onChange} id="password" />);
		tree.find('TextInput').simulate('change', { target: { placeholder: 'password', value: 'secret' } });
		expect(onChange).toBeCalledWith(
			expect.objectContaining({
				error: 'password should at least be 8 characters',
				isValid: false,
			}),
		);
	});

	it('should identify weak passwords', () => {
		const onChange = jest.fn();
		const tree = shallow(<PasswordInput onChange={onChange} id="password" showPasswordStrength />);
		tree.find('TextInput').simulate('change', { target: { placeholder: 'password', value: 'secretttt' } });
		expect(onChange).toBeCalledWith(
			expect.objectContaining({
				passwordStrength: 'weak',
			}),
		);
	});

	it('should identify medium passwords', () => {
		const onChange = jest.fn();
		const tree = shallow(<PasswordInput onChange={onChange} id="password" showPasswordStrength />);
		tree.find('TextInput').simulate('change', { target: { placeholder: 'password', value: 'secret1!@#2' } });
		expect(onChange).toBeCalledWith(
			expect.objectContaining({
				passwordStrength: 'medium',
			}),
		);
	});

	it('should identify strong passwords', () => {
		const onChange = jest.fn();
		const tree = shallow(<PasswordInput onChange={onChange} id="password" showPasswordStrength />);
		tree.find('TextInput').simulate('change', { target: { placeholder: 'password', value: 'secret123!@#123' } });
		expect(onChange).toBeCalledWith(
			expect.objectContaining({
				passwordStrength: 'strong',
			}),
		);
	});

	it('should flag empty passwords', () => {
		const onChange = jest.fn();
		const tree = shallow(<PasswordInput onChange={onChange} id="password" />);
		tree.find('TextInput').simulate('change', { target: { value: '' } });
		expect(onChange).toBeCalledWith(
			expect.objectContaining({
				error: 'Password can not be empty',
				isValid: false,
			}),
		);
	});

	it('should show password visibility', () => {
		const onChange = jest.fn();
		const handleVisibilityStatus = jest.fn();
		const tree = shallow(
			<PasswordInput
				onChange={onChange}
				id="password"
				showPasswordStrength
				shouldShowVisibility
				handleVisibilityStatus={handleVisibilityStatus}
			/>
		);
		tree.find('TextInput').simulate('changePasswordVisibilityStatus');
		expect(handleVisibilityStatus).toBeCalled();
	});
});
