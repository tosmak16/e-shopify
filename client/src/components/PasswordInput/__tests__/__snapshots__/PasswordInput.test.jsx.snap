// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput renders 1`] = `
<div>
  <TextInput
    className=""
    dark={false}
    disabled={false}
    errorMessage=""
    handleVisibilityStatus={[Function]}
    id="password"
    inputDidChange={false}
    isValid={true}
    label=""
    name=""
    onBlur={null}
    onChange={[Function]}
    onChangePasswordVisibilityStatus={[Function]}
    onKeyDown={[Function]}
    placeholder=""
    shouldShowVisibility={false}
    showPasswordStrength={false}
    size={8}
    type="password"
    value=""
  />
</div>
`;
